design/00292: EEPROM storage.md

```diff
@@ -0,0 +1,64 @@
+
+## Abstract
+
+The following document proposes to store the PlanktoScope (PS) hardware configuration in the 32 Kbits EEPROM chip of the PlanktoScope HAT in order to : 
+* automatically save the hardware state of the machine
+* easily retrieve the hardware configuration for debugging
+* feed the metadata with precise hardware information
+
+## Background
+
+A specific HAT was developped to satisfy the PlanktoScope needs. 
+This PlanktoScope HAT connected to the raspberry pi and all of the external devices (led, motors, pump, camera) contains a 32 Kbit EEPROM chip which is a "type of non-volatile ROM that enables individual bytes of data to be erased and reprogrammed" (Wikipedia).
+It is therefore possible to store some data inside this chip and to read it back.
+
+For now (Sunray version 2024.0.0-beta.2) some of the hardware configuration of the PlanktoScope is entered by the user in the GUI.
+We would like to have an hardware configuration specific to each machine in case some specific modification is made.
+We would like to store this configuration in a safe place, independent from the software installed on the raspberry pi.
+
+## Proposal
+
+We therefore propose to write a python script that allows the storage and the reading of the PlanktoScope hardware information on the EEPROM chip.
+
+This information stored should contain : 
+* PlanktoScope reference
+* Plantoscope serial number
+* PlanktoScope version
+* Plantoscope fabrication date
+* HAT serial number
+* HAT version
+* Driver reference
+* Pump reference
+* Stepper focus reference
+* Objective lens reference
+* Tube lens reference
+* Flowcell thickness
+* LED reference
+
+Those informations could be used to feed the metadata stored in the hardware.json file as described in the issue 290 (https://github.com/PlanktoScope/PlanktoScope/issues/290).
+
+## Rationale
+
+The EEPROM chip currently used for the HAT has a capacity of 32Kbits. This provides adequate space for the necessary configuration without redundancy or excessive overhead.
+The use of our HAT is relevant as the PlanktoScope HAT is linked to every device of the machine. This allows unique data storage for each PlanktoScope unit, helping track individual hardware changes or updates.
+The EEPROM on the HAT is accessible directly through the I2C bus of the Raspberry Pi, simplifying software development.
+
+This solution is exclusive to the PlanktoScope HAT, limiting compatibility with other configurations (e.g., Adafruit HATs without EEPROM).
+Though currently sufficient, 32Kbits could become a constraint if future features or expanded data storage needs arise.
+
+## Compatibility
+
+The main compatibility issue concerns the adafruit HAT users.
+For those, a simple solution would be to continue selecting their configuration in the GUI, or adapt the script to store informations in the Raspberry pi EEPROM chip in case of not finding the PlanktoScope HAT EEPROM adress.
```

Comment:
Based on https://www.raspberrypi.com/documentation/computers/raspberry-pi.html , I believe the Raspberry Pi's EEPROM chip may already be used/reserved by the Raspberry Pi as part of how it boots up. If this is a correct understanding of the Raspberry Pi, then we cannot use the Raspberry Pi's EEPROM chip to store our own data.

When there is no PlanktoScope HAT with EEPROM, the other alternative would be to use a JSON file on the SD card as a substitute for storing data on EEPROM. I recommend you to describe this option in this design document.